pkg/components/util/delete.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"strings"
+
+	"helm.sh/helm/v3/pkg/action"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kinvolk/lokomotive/pkg/components"
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+)
+
+// DeleteComponent deletes a component.
+func DeleteComponent(name string, c components.Component, kubeconfig string, deleteNS bool) error {
+	ns := c.Metadata().Namespace
+	if ns == "" {
+		return fmt.Errorf("component %s namespace is empty", name)
+	}
+
+	cfg, err := HelmActionConfig(ns, kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed preparing helm client: %w", err)
+	}
+
+	var errs []string
+
+	uninstall := action.NewUninstall(cfg)
+
+	if resp, err := uninstall.Run(name); err != nil {
+		errs = append(errs, err.Error())
+	} else {
+		fmt.Println(resp.Info)
+	}
+
+	if deleteNS {
+		if err := deleteNamespace(ns, kubeconfig); err != nil {
+			errs = append(errs, err.Error())
+		}
+	}
+
+	return AggregateErrors(errs)
+}
+
+func deleteNamespace(ns string, kubeconfig string) error {
+	cs, err := k8sutil.NewClientset(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	// Delete the manually created namespace which was not created by helm.
+	if err = cs.CoreV1().Namespaces().Delete(ns, &metav1.DeleteOptions{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AggregateErrors conflates multiple errors stored in a string form and returns a single error.
+// nolint:gomnd
+func AggregateErrors(errs []string) error {
```

Comment:
Do we really want to continue uninstalling components on error? IMO it may be safer to stop the execution on the first error we encounter. If the command is idempotent, the user can fix the error and then re-run the command to "resume" the deletion.

If we do this ^ then we no longer need this function.

Reply:
> Do we really want to continue uninstalling components on error? IMO it may be safer to stop the execution on the first error we encounter. If the command is idempotent, the user can fix the error and then re-run the command to "resume" the deletion.
> 
> If we do this ^ then we no longer need this function.


I think error in deletion of one component should not hinder deletion of other components.

---

That being said here is how the functionality will look like if we return immediately. In the case below, one cannot fix the error because you cannot delete namespace.

```console
$ lokoctl component delete --delete-namespace
Do you want to delete components: "metrics-server,external-dns"? [type "yes" to continue]: yes
Deleting component 'metrics-server'...

FATA[0008] namespaces "kube-system" is forbidden: this namespace may not be deleted  args="[]" command="lokoctl component delete"
```

You could possibly omit the flag `--delete-namespace` and things will be ok. But not really since the order of deletion cannot be changed. 

```console
$ lokoctl component delete
Do you want to delete components: "metrics-server,external-dns"? [type "yes" to continue]: yes
Deleting component 'metrics-server'...
FATA[0003] uninstall: Release not loaded: metrics-server: release: not found  args="[]" command="lokoctl component delete"You could possibly not provide the flag `--delete-namespace` and things will be ok.
```

Now one can provide the component name explicitly to go forward.

```console
$ lokoctl component delete external-dns --delete-namespace
Do you want to delete components: "external-dns"? [type "yes" to continue]: yes
Deleting component 'external-dns'...

Successfully deleted component "external-dns"!
```

---


But in current implementation, here you can see that error in deletion of one component did not hinder deletion of others.

```console
$ lokoctl component delete --delete-namespace
Do you want to delete components: "metrics-server,external-dns"? [type "yes" to continue]: yes
Deleting component 'metrics-server'...

Deleting component 'external-dns'...

Successfully deleted component "external-dns"!

FATA[0014] namespaces "kube-system" is forbidden: this namespace may not be deleted  args="[]" command="lokoctl component delete"
```

Reply:
Ideally if the release is not found this should not be an error and we should continue to make it idempotent. Is that something we can do?

Reply:
> Ideally if the release is not found this should not be an error and we should continue to make it idempotent. Is that something we can do?

Added a check which ignores that error.